```astro:env``` breaks in Content Collection ```config.ts```

### Astro Info

```block
Astro                    v4.14.2
Node                     v20.14.0
System                   macOS (arm64)
Package Manager          npm
Output                   server
Adapter                  @astrojs/node
Integrations             @astrojs/tailwind
                         @astrojs/solid-js
                         @astrojs/react
```


### If this issue only occurs in one browser, which browser is a problem?

_No response_

### Describe the Bug

Astro environments are not working in the configuration for Content Collections ```config.ts```. It seems that the generation code for the Content Collections is running before the generation code for the environment variables runs, therefore the content collection generation process is breaking? It seems that the environments needs to be fully generated first before the content collections generate?

```
[ERROR] [types] [GenerateContentTypesError] `astro sync` command failed to generate content collection types: Cannot find module 'astro:env/client' imported from './src/content/config.ts'
```

```
Stack trace:
    at syncContentCollections (file:///.../node_modules/astro/dist/core/sync/index.js:156:11)
    at async sync (file:///.../node_modules/astro/dist/core/sync/index.js:45:10)
    at async runCommand (file:///.../node_modules/astro/dist/cli/index.js:153:27)
  Caused by:
  Cannot find module 'astro:env/client' imported from '.../src/content/config.ts'
    at nodeImport (file:///.../node_modules/vite/dist/node/chunks/dep-NjL7WTE1.js:52919:19)
    at eval (.../src/content/config.ts:5:37)
```

### What's the expected result?

Environment variables are useful for dynamically configuring where content should be loaded from, and the astro:env module should be accessible for the code based content configuration

### Link to Minimal Reproducible Example

pending

### Participation

- [ ] I am willing to submit a pull request for this issue.

--- a/.changeset/tiny-lamps-lick.md
+++ b/.changeset/tiny-lamps-lick.md
@@ -1,0 +1,5 @@
+---
+'astro': patch
+---
+
+Fixes an error thrown by `astro sync` when an `astro:env` virtual module is imported inside the Content Collections config

--- a/packages/astro/src/env/vite-plugin-env.ts
+++ b/packages/astro/src/env/vite-plugin-env.ts
@@ -31,7 +31,7 @@ export function astroEnv({
 	fs,
 	sync,
 }: AstroEnvVirtualModPluginParams): Plugin | undefined {
-	if (!settings.config.experimental.env || sync) {
+	if (!settings.config.experimental.env) {
 		return;
 	}
 	const schema = settings.config.experimental.env.schema ?? {};
@@ -57,6 +57,7 @@ export function astroEnv({
 				schema,
 				loadedEnv,
 				validateSecrets: settings.config.experimental.env?.validateSecrets ?? false,
+				sync,
 			});
 
 			templates = {
@@ -100,10 +101,12 @@ function validatePublicVariables({
 	schema,
 	loadedEnv,
 	validateSecrets,
+	sync,
 }: {
 	schema: EnvSchema;
 	loadedEnv: Record<string, string>;
 	validateSecrets: boolean;
+	sync: boolean;
 }) {
 	const valid: Array<{ key: string; value: any; type: string; context: 'server' | 'client' }> = [];
 	const invalid: Array<InvalidVariable> = [];
@@ -125,7 +128,7 @@ function validatePublicVariables({
 		}
 	}
 
-	if (invalid.length > 0) {
+	if (invalid.length > 0 && !sync) {
 		throw new AstroError({
 			...AstroErrorData.EnvInvalidVariables,
 			message: AstroErrorData.EnvInvalidVariables.message(invalidVariablesToError(invalid)),

--- a/packages/astro/test/astro-sync.test.js
+++ b/packages/astro/test/astro-sync.test.js
@@ -219,6 +219,16 @@ describe('astro sync', () => {
 				assert.fail();
 			}
 		});
+		it('Does not throw if a virtual module is imported in content/config.ts', async () => {
+			try {
+				await fixture.load('./fixtures/astro-env-content-collections/');
+				fixture.clean();
+				await fixture.whenSyncing();
+				assert.ok(true);
+			} catch {
+				assert.fail();
+			}
+		});
 	});
 
 	describe('astro:actions', () => {

--- a/packages/astro/test/fixtures/astro-env-content-collections/astro.config.mjs
+++ b/packages/astro/test/fixtures/astro-env-content-collections/astro.config.mjs
@@ -1,0 +1,12 @@
+import { defineConfig, envField } from 'astro/config';
+
+// https://astro.build/config
+export default defineConfig({
+	experimental: {
+		env: {
+			schema: {
+				FOO: envField.string({ context: "client", access: "public", optional: true, default: "ABC" }),
+			}
+		}
+	}
+});

--- a/packages/astro/test/fixtures/astro-env-content-collections/package.json
+++ b/packages/astro/test/fixtures/astro-env-content-collections/package.json
@@ -1,0 +1,8 @@
+{
+  "name": "@test/astro-env-content-collections",
+  "version": "0.0.0",
+  "private": true,
+  "dependencies": {
+    "astro": "workspace:*"
+  }
+}

--- a/packages/astro/test/fixtures/astro-env-content-collections/src/content/config.ts
+++ b/packages/astro/test/fixtures/astro-env-content-collections/src/content/config.ts
@@ -1,0 +1,13 @@
+import { defineCollection, z } from "astro:content";
+import { FOO } from "astro:env/client"
+
+console.log({ FOO })
+
+export const collections = {
+    foo: defineCollection({
+        type: "data",
+        schema: z.object({
+            title: z.string()
+        })
+    })
+}

--- a/packages/astro/test/fixtures/astro-env-content-collections/tsconfig.json
+++ b/packages/astro/test/fixtures/astro-env-content-collections/tsconfig.json
@@ -1,0 +1,3 @@
+{
+  "extends": "astro/tsconfigs/base"
+}

